test/shopify_app/auth/token_exchange_test.rb

```diff
@@ -0,0 +1,177 @@
+# frozen_string_literal: true
+
+require "test_helper"
+
+class ShopifyApp::Auth::TokenExchangeTest < ActiveSupport::TestCase
+  OFFLINE_ACCESS_TOKEN_TYPE = ShopifyAPI::Auth::TokenExchange::RequestedTokenType::OFFLINE_ACCESS_TOKEN
+  ONLINE_ACCESS_TOKEN_TYPE = ShopifyAPI::Auth::TokenExchange::RequestedTokenType::ONLINE_ACCESS_TOKEN
+
+  def setup
+    ShopifyApp::SessionRepository.shop_storage = ShopifyApp::InMemoryShopSessionStore
+    ShopifyApp::SessionRepository.user_storage = nil
+
+    @shop = "my-shop.myshopify.com"
+    @user_id = 1
+    @session_token = build_jwt
+
+    @offline_session = build_offline_session
+    @online_session = build_online_session
+
+    ShopifyApp.configuration.post_authenticate_tasks.stubs(:perform)
+  end
+
+  test "#perform exchanges offline token then stores it when only shop session store is configured" do
+    ShopifyAPI::Auth::TokenExchange.expects(:exchange_token).with(
+      shop: @shop,
+      session_token: @session_token,
+      requested_token_type: OFFLINE_ACCESS_TOKEN_TYPE,
+    ).once.returns(@offline_session)
+
+    assert_nil ShopifyApp::SessionRepository.load_session(@offline_session.id)
+
+    new_session = ShopifyApp::Auth::TokenExchange.perform(@session_token)
+
+    assert_equal @offline_session, new_session
+    assert_equal @offline_session, ShopifyApp::SessionRepository.load_session(@offline_session.id)
+  end
+
+  test "#perform exchanges both online and offline tokens then stores it when user session store is configured" do
+    ShopifyApp::SessionRepository.user_storage = ShopifyApp::InMemoryUserSessionStore
+
+    ShopifyAPI::Auth::TokenExchange.expects(:exchange_token).with(
+      shop: @shop,
+      session_token: @session_token,
+      requested_token_type: OFFLINE_ACCESS_TOKEN_TYPE,
+    ).returns(@offline_session)
+
+    ShopifyAPI::Auth::TokenExchange.expects(:exchange_token).with(
+      shop: @shop,
+      session_token: @session_token,
+      requested_token_type: ONLINE_ACCESS_TOKEN_TYPE,
+    ).returns(@online_session)
+
+    assert_nil ShopifyApp::SessionRepository.load_session(@online_session.id)
```

Comment:
Nit: should we check that the offline session _also_ got stored here?

Reply:
Came here to say the same thing :)